### ISSUE TITLE
[QUESTION]: What is the difference between ts in key and version in meta

### ISSUE THREAD

**Issue by guoxiangCN** [dgraph-io/badger#1958]
### Question.

Hello, I am reading the source code, but I have some questions. Can someone help me answer the following? Thank you very much!
Question1. What is the difference between ts in key and version in meta?
Question2. Why is it that when searching for a key, the closest readTs is found in the memtable, but still needs to be searched in different levels?

**Comment by MichelDiz** [comment]
A little wordy, but here it goes.

Question 1: The ts in the key and the version in the meta are essentially the same thing. They both represent the version of the key. In Badger, each key can have multiple versions, each with a different value. The version is stored as part of the key in the LSM tree, but it's also stored in the value log (if the value is large enough to be written to the value log). The ts in the key is used for versioning in the LSM tree, while the version in the meta is used for versioning in the value log.

Question 2: The reason for this is because Badger is an LSM (Log-Structured Merge) tree-based database. In an LSM tree, data is first written to a memory table (memtable). When the memtable is full, it's flushed to disk as a sorted string table (SSTable) at level 0. Over time, SSTables are compacted and moved to higher levels. Each level can contain multiple SSTables, and each SSTable can contain multiple keys. Therefore, even if a key is found in the memtable, it's still necessary to search the SSTables in different levels to find all versions of the key. This is because older versions of the key could be in SSTables at higher levels.

Check this video to learn more about LSM trees

https://www.youtube.com/watch?v=I6jB0nM9SKU&feature

**Comment by guoxiangCN** [comment]
Thanks for your reply ! I still have some questions
1. the version will be set to commit-ts before the txn commits, but after it was persist in the valuelog, the version will be set to zero and then insert into the the lsm-tree(wal+memtale); however, when the vlog do the gc and rewrites, the version also be set to zero and then insert into the new vlog+wal+memtables. These scenarios confuse me with the value of version。

2. for the question 2 above, I have read the code for leveldb/rocksdb before, their query logic will be immediately interrupted when get the latest version of a key. For example, when the value of a key is found in memtable, it will immediately return without the need to continue searching for memtable and sst. If it is found at level 0, it will not search for levels 1 or higher.
But when I read Badger's code, I found that even if I found the value of a key in memtable or low level,  still needed to search for each level, which made me confused and I didn't quite understand the original intention.
